src/share/classes/com/sun/javatest/regtest/agent/JUnitRunner.java

```diff
@@ -126,6 +133,7 @@ private static void runWithJUnitPlatform(Class<?> mainClass) throws Exception {
             try (LauncherSession session = LauncherFactory.openSession()) {
                 Launcher launcher = session.getLauncher();
                 launcher.registerTestExecutionListeners(summaryGeneratingListener);
+                launcher.registerTestExecutionListeners(new PrintingListener(System.err));
```

Comment:
Is System.err correct? If my test prints messages to System.out them how will I know which test they came from?

Reply:
`System.err` seems to be more correct, looking at other printing code in jtreg. I used `System.out` first, which was remarked by @jonathan-gibbons in a previous note.

> If my test prints messages to System.out them how will I know which test they came from?

The canonical way is Jupiter's `TestReporter` as described in [Writing Tests - Dependency Injection for Constructors and Methods](https://junit.org/junit5/docs/current/user-guide/#writing-tests-dependency-injection)

> The `TestReporter` can be used to publish additional data about the current test run. The data can be consumed via the `reportingEntryPublished()` method in a [TestExecutionListener](https://junit.org/junit5/docs/current/api/org.junit.platform.launcher/org/junit/platform/launcher/TestExecutionListener.html), allowing it to be viewed in IDEs or included in reports.

```
class TestReporterDemo {

    @Test
    void reportSingleValue(TestReporter testReporter) {
        testReporter.publishEntry("a status message");
    }

    @Test
    void reportKeyValuePair(TestReporter testReporter) {
        testReporter.publishEntry("a key", "a value");
    }

    @Test
    void reportMultipleKeyValuePairs(TestReporter testReporter) {
        var values = Map.of("user name", "dk38", "award year", "1974");
        testReporter.publishEntry(values);
    }

}
```

Reply:
Having said that, there's also: [Capturing Standard Output/Error](https://junit.org/junit5/docs/current/user-guide/#running-tests-capturing-output), an opt-in support for capturing output printed to `System.out` and `System.err`.
> If enabled, the JUnit Platform captures the corresponding output and publishes it as a report entry using the stdout or stderr keys to all registered [TestExecutionListener](https://junit.org/junit5/docs/current/api/org.junit.platform.launcher/org/junit/platform/launcher/TestExecutionListener.html) instances immediately before reporting the test or container as finished.

Shall we enable capturing standard streams by default?

Reply:
With TestNG, the "test $testname: $result" messages show up in the jtr System.out section. If a test prints anything then it will inlined and appear before the message emitted by TestNG. So this will take a bit of getting used to.

Reply:
It's just a matter of remembering that testng uses System.out while jtreg uses System.err. If you use logging in your tests (either directly or by enabling logging for the lib you're using - e.g. http traces) then these traces will go to System.err. In some cases I had to go out of my way to configure the logging to go to System.out instead when using testng. So System.err is probably less surprising unless you come from testng ;-)

Reply:
See https://en.wikipedia.org/wiki/Standard_streams
Generally, `stdout` is normally reserved for output relating to the function of a tool, like command-line help, and `tier` is used for diagnostics and logging messages.